openmp/libomptarget/test/libc/fwrite.c

```diff
@@ -0,0 +1,22 @@
+// RUN: %libomptarget-compile-run-and-check-generic
+
+// REQUIRES: libc
+
+#include <assert.h>
+#include <stdio.h>
+
+#pragma omp declare target to(stdout)
+
+int main() {
+  int r = 0;
+// CHECK: PASS
+#pragma omp target map(from : r)
+  { r = fwrite("PASS\n", 1, sizeof("PASS\n") - 1, stdout); }
+  assert(r == sizeof("PASS\n") - 1 && "Incorrect number of bytes written");
```

Comment:
Could you explain what thread/action uses which stream here?
I'm not 100% sure the extra sync above is sufficient (e.g., what about H2D?) and necessary.
Maybe we should just have more streams involved, but I am not clear what streams we currently use (for the RPC parts).

Reply:
So, as far as I understand it, for a target region with a copy-back, we simply call this function to retrieve the data. Currently, NVPTX does an implicit sync because we schedule both the kernel and the copy-back on the steam stream. i.e. we cannot issue the memcpy until the kernel is done so we wait inside of the CUDA runtime. This is bad because it means that we aren't periodicaly checking the RPC server while waiting.

Reply:
> we schedule both the kernel and the copy-back on the steam stream.
Yes, that is correct.

> i.e. we cannot issue the memcpy until the kernel is done so we wait inside of the CUDA runtime.
That I don't get.

> This is bad because it means that we aren't periodicaly checking the RPC server while waiting.
The checking is done by the thread that issued the kernel, right? It does not wait for the stream with the kernel to finish before it checks the RPC stream, correct? I don't get why this interferes. I would assume it checks the RCP, then non-blocking the kernel, and so on?

Reply:
If you go into a debugger, it waits forever inside of the `D2H` memcpy function. This is because CUDA is most likely waiting internally for the kernel to be done. THat means no one on our side is checking the server.

Reply:
> If you go into a debugger, it waits forever inside of the `D2H` memcpy function. This is because CUDA is most likely waiting internally for the kernel to be done. THat means no one on our side is checking the server.

I don't disagree with this but it does not answer my questions above. I can try to look into the current state next week myself.

Reply:
The current behavior looks like all those async operations can only be "enqueued" at certain point but because the device is busy waiting for the RPC request, it didn't reach the point. As a consequence, the host side is stuck at "issuing" stage.

Reply:
Yeah, this could be solved by forking a thread for the RPC server to run in the background, but it seemed better to re-use waiting threads we already have. Changing that is another alternative if we wanted to just fork another "hidden" thread.

Reply:
So, turns out the call is blocking, according to https://docs.nvidia.com/cuda/cuda-driver-api/api-sync-behavior.html#api-sync-behavior__memcpy-async rule synchronous 3.
The solution above seems sensible to me.